File: find_Ngrams.py
from nltk.util import ngrams
from collections import Counter
from joblib import Parallel, delayed
import asyncio


# Fonction pour générer les n-grams
def generate_ngrams(tokens, n):
    return [' '.join(gram) for gram in ngrams(tokens, n)]


# Fonction pour compter les occurrences des n-grams
def count_ngrams(ngrams_list):
    return Counter(ngrams_list)


# Fonction pour générer les n-grams en parallèle avec joblib
def generate_ngrams_joblib(tokens, n, n_jobs=4):
    chunk_size = len(tokens) // n_jobs
    chunks = [tokens[i:i + chunk_size] for i in range(0, len(tokens), chunk_size)]
    results = Parallel(n_jobs=n_jobs)(delayed(generate_ngrams)(chunk, n) for chunk in chunks)
    ngrams_list = [ngram for sublist in results for ngram in sublist]
    return ngrams_list


# Fonction pour générer les n-grams en parallèle avec asyncio
async def generate_ngrams_asyncio(tokens, n, chunk_size):
    async def process_chunk(chunk):
        return generate_ngrams(chunk, n)

    chunks = [tokens[i:i + chunk_size] for i in range(0, len(tokens), chunk_size)]
    tasks = [process_chunk(chunk) for chunk in chunks]
    results = await asyncio.gather(*tasks)
    ngrams_list = [ngram for sublist in results for ngram in sublist]
    return ngrams_list

File: find_Ngrams.cpp
//
// Created by dylan on 01/07/24.
//

#include "find_Ngrams.h"


// Fonction pour générer les n-grams
std::vector<std::string> generate_ngrams_par(const std::vector<std::string>& tokens, int n) {
    std::vector<std::string> ngrams;
    int size = tokens.size();

#pragma omp parallel for
    for (int i = 0; i <= size - n; ++i) {
        std::string ngram = tokens[i];
        for (int j = 1; j < n; ++j) {
            ngram += " " + tokens[i + j];
        }
#pragma omp critical
        ngrams.push_back(ngram);
    }

    return ngrams;
}

// Fonction pour générer les n-grams
std::vector<std::string> generate_ngrams(const std::vector<std::string>& tokens, int n) {
    std::vector<std::string> ngrams;
    int size = tokens.size();

    for (int i = 0; i <= size - n; ++i) {
        std::string ngram = tokens[i];
        for (int j = 1; j < n; ++j) {
            ngram += " " + tokens[i + j];
        }
        ngrams.push_back(ngram);
    }

    return ngrams;
}

// Fonction pour compter les occurrences des n-grams
std::unordered_map<std::string, int> count_ngrams_par(const std::vector<std::string>& ngrams) {
    std::unordered_map<std::string, int> ngram_counts;

#pragma omp parallel for
    for (size_t i = 0; i < ngrams.size(); ++i) {
#pragma omp critical
        ngram_counts[ngrams[i]]++;
    }

    return ngram_counts;
}

// Fonction pour compter les occurrences des n-grams
std::unordered_map<std::string, int> count_ngrams(const std::vector<std::string>& ngrams) {
    std::unordered_map<std::string, int> ngram_counts;

    for (size_t i = 0; i < ngrams.size(); ++i) {
        ngram_counts[ngrams[i]]++;
    }

    return ngram_counts;
}
